ui/app/ducks/swaps/swaps.js

```diff
@@ -536,3 +592,36 @@ export const signAndSendTransactions = (history, metaMetricsEvent) => {
     await forceUpdateMetamaskState(dispatch)
   }
 }
+
+export function fetchMetaSwapsGasPriceEstimates () {
+  return async (dispatch, getState) => {
+    const state = getState()
+    const priceEstimatesLastRetrieved = getSwapsPriceEstimatesLastRetrieved(state)
+    const timeLastRetrieved = priceEstimatesLastRetrieved || loadLocalStorageData('METASWAP_GAS_PRICE_ESTIMATES_LAST_RETRIEVED') || 0
+
+    dispatch(setSwapGasPriceEstimatesLoadingState(true))
+
+    let priceEstimates
+    if (Date.now() - timeLastRetrieved > 30000) {
+      priceEstimates = await fetchExternalMetaSwapGasPriceEstimates(dispatch)
+    } else {
+      const cachedPriceEstimates = loadLocalStorageData('METASWAP_GAS_PRICE_ESTIMATES')
+      priceEstimates = cachedPriceEstimates || await fetchExternalMetaSwapGasPriceEstimates(dispatch)
+    }
+
+    dispatch(setSwapGasPriceEstimates(priceEstimates))
+    dispatch(setSwapGasPriceEstimatesLoadingState(false))
+    return priceEstimates
+  }
+}
+
+async function fetchExternalMetaSwapGasPriceEstimates (dispatch) {
+  const priceEstimates = await fetchSwapsGasPrices()
+
+  const timeRetrieved = Date.now()
+  saveLocalStorageData(priceEstimates, 'METASWAP_GAS_PRICE_ESTIMATES')
+  saveLocalStorageData(timeRetrieved, 'METASWAP_GAS_PRICE_ESTIMATES_LAST_RETRIEVED')
+  dispatch(setSwapsGasPriceEstimatesLastRetrieved(timeRetrieved))
```

Comment:
This dispatch can be combined with the event mentioned here https://github.com/MetaMask/metamask-extension/pull/9599/files#r510256433

Reply:
4c8f00846